dpc/src/record/encoded.rs

```diff
@@ -16,45 +16,90 @@
+    let capacity = <F::Parameters as FieldParameters>::CAPACITY as usize;
+
+    let mut bits = Vec::<bool>::with_capacity(x_bytes.len() * 8 + 1);
+    for byte in x_bytes.iter() {
+        let mut byte = byte.clone();
+        for _ in 0..8 {
+            bits.push(byte & 1 == 1);
+            byte >>= 1;
+        }
+    }
+    bits.push(true);
+    // The last bit indicates the end of the actual data, which is used in decoding to
+    // make sure that the length is correct.
 
-use itertools::Itertools;
-use std::marker::PhantomData;
-
-/// Encode a base field element bytes to a group representation
-pub fn encode_to_group<P: MontgomeryParameters + TwistedEdwardsParameters, G: ProjectiveCurve>(
-    x_bytes: &[u8],
-) -> Result<(<G as ProjectiveCurve>::Affine, bool), DPCError> {
-    // TODO (howardwu): Remove this hardcoded value and use BaseField's size in bits to pad length.
-    let mut bytes = x_bytes.to_vec();
-    while bytes.len() < 32 {
-        bytes.push(0)
+    let mut res = Vec::<F>::with_capacity((bits.len() + capacity - 1) / capacity);
+    for chunk in bits.chunks(capacity) {
+        res.push(F::from_repr(F::BigInteger::from_bits_le(chunk)).unwrap());
     }
 
-    let input = P::BaseField::read_le(&bytes[..])?;
-    let (output, fq_high) = Elligator2::<P, G>::encode(&input)?;
-    Ok((output, fq_high))
+    Ok(res)
 }
 
-/// Decode a group into the byte representation of a base field element
-pub fn decode_from_group<P: MontgomeryParameters + TwistedEdwardsParameters, G: ProjectiveCurve>(
-    affine: <G as ProjectiveCurve>::Affine,
-    fq_high: bool,
-) -> Result<Vec<u8>, DPCError> {
-    let output = Elligator2::<P, G>::decode(&affine, fq_high)?;
-    Ok(to_bytes_le![output]?)
+/// Decode a group into the byte representation of a base field element.
+pub fn decode_from_field<F: PrimeField>(field_elements: &[F]) -> Result<Vec<u8>, DPCError> {
+    if field_elements.len() == 0 {
+        return Err(DPCError::EncodingError(EncodingError::Message(
+            "The encoded record must consist of at least one field element.".to_string(),
+        )));
+    }
+    // There is at least one field element due to the additional true bit.
+
+    let capacity = <F::Parameters as FieldParameters>::CAPACITY as usize;
+
+    let mut bits = Vec::<bool>::with_capacity(field_elements.len() * capacity);
+    for elem in field_elements.iter() {
+        let elem_bits = elem.to_repr().to_bits_le();
+        bits.extend_from_slice(&elem_bits[0..capacity]); // only keep `capacity` bits, discarding the highest bit.
+    }
+
+    // Find the last bit of `true`, which should be within the last `CAPACITY` bits.
+    // If it does not exist, this is a `DPCError::EncodingError`.
+    let mut last_bit_pos = None;
+    for i in ((bits.len() - capacity)..(bits.len())).rev() {
+        if bits[i] == true {
+            last_bit_pos = Some(i);
+            break;
+        }
+    }
```

Comment:
To make this more readable, I believe you can do the following:
```rs
let mut found_last_bit = false;
while !found_last_bit {
    if let Some(true) = bits.pop() {
        found_last_bit = true;
    }
}
```

This forgoes needing to keep a track of the index, and allows you to remove the `bits.truncate()` call after.

Remember to account for the difference in `.rev()` in the final for loop.

Reply:
This is better than my suggestion, considering `bits` is being truncated later on :+1:.

Reply:
Lukasz just provided a way to rewrite it. Let me use that.

I limit the amount of bits to be considered to be <= `capacity` to ensure that an overpadding ciphertext is rejected.

Reply:
(here by overpadding I mean a ciphertext whose last "1" bit is not in the last field element of the encoding, but in a prior one.)

Reply:
If the `capacity` is something you wish to enforce, I suggest a different approach.

`(0..bits.len()).rev().take(capacity)` will not enforce `capacity`. Notice it will take up to `bits.len()` and index directly into that. What I think you want is to `take` from `bits` itself then.

Otherwise @ljedrz approach and my approach functionally will fail under the same conditions.

Reply:
In fact, it may be better to explicitly check the length is `>= capacity` first, and then use an index-less while loop.

Reply:
There's also another longish option that avoids indexing:
```
let last_bit_pos = bits.iter().enumerate().rev().take(capacity).find(|(i_, &b)| b).map(|(i, _b)| i);
```

Reply:
and yes, `explicitly check the length is >= capacity first` is done, as it requires the plaintext to have at least one element.

Reply:
^That does work to get the index, however I believe it still does not enforce `bits.len >= capacity`. `take(capacity)` will still "fail gracefully" and `find` won't catch the issue.

Reply:
It is at least equal, as `field_elements` is checked to be non-empty, and `bits` gets populated with `field_elements.len() * capacity`.

Reply:
I think there is a guarantee: (let me correct my sentence).

the decode algorithm requires the encoded record to have at least one field element
```
    if field_elements.is_empty() {
        return Err(EncodingError::Message(
            "The encoded record must consist of at least one field element.".to_string(),
        )
        .into());
    }
```

and as a result, the minimal len of `bits` is capacity.

Reply:
Here is a simple example to demonstrate `capacity` isn't enforced.

```rs
fn main() {
    let bits = [true, false, true];
    let capacity = 5;
    
    let last_bit_pos = bits.iter().enumerate().rev().take(capacity).find(|(i_, &b)| b).map(|(i, _b)| i);
    assert_eq!(Some(2), last_bit_pos);
}
```

Reply:
Indeed, it is not enforced by the iterator itself; that one only avoids indexing, which is a good thing in general, not just for security but often performance as well.

Reply:
Then I think this could work as well (just a revision on my original approach):
```rs
loop {
    if let Some(true) = bits.pop() {
        break;
    }
}
```

Reply:
That one will go over `capacity`, though; `loop` would need to be `for _ in 0..capacity` instead to limit the maximum number of iterations, if the original behavior is to be preserved.

Reply:
Spoke with @weikengchen offline, we now have a better picture of the issue.

TLDR - we need the last field element from the given array to be zero. A simple check of that will remove any of this capacity checks and simplify the code significantly.

Reply:
The logic has been optimized in https://github.com/AleoHQ/snarkVM/pull/311/commits/dbc9dffe2c17e5544c093dec50cbfab780766cba

Reply:
LGTM

Reply:
Wait, actually, don't use the while loop. I take back my suggestion.

That while loop will infinite loop if given an array of all false's.

Reply:
And it has been simplified again since we now enforce the last element to be nonzero, everything is simpler.

Reply:
Okay cool, you caught the issue I was gonna call out, which is the error case after is no longer needed.